### PR TITLE
Fix a set of E2E bugs causing several CI failures

### LINKED ISSUE #2175: E2E tests are not working in development
A summary of the latest run I made for **e2e** tests under localhost. 
There are a bunch of tests that are failing. I suspect this is true also for CI based on the list of failures within the open PR pool. 

image

### Steps to Reproduce

_as it is instructed by the README_ 
1. `yarn dx` in one terminal 
2. `yarn workspace @calcom/web test-e2e`  in another terminal

### PR DESCRIPTION
**Pull request by denik1981**
## What does this PR do?

This is a group of several different fixes against the E2E tests running with Playwright. 
The following summarizes each of the fixes made here:

**WORKERS = 1** _(Playwright Configuration)_
Based on the way we are currently running tests, following concurrency can lead us to corrupted tests. A commit into the DB made by one of the test threads can get into conflict with another thread looking for the same data. To reestablish parallelism my suggestion is to introduce **Fixtures** in our tests and scope each of them to newly created data. More about this at the bottom.   

**RETRY=1**  _(Playwright Configuration)_
A lifesaver at least when running tests locally. I can confirm that I have executed several tests that fail in the first run but succeed in the retry.

**NEXT_PUBLIC_IS_E2E**
There is one failing test using a custom header [x-playwright-secret] that is looking to match our custom PLAYWRIGHT_SECRET env variable. This is now replaced with a global (client and server) E2E env that will allow us have more control in terms of how to bypass certain logic in the codebase when running under E2E. This was the main cause of the  `forgot-password` test to fail.

**PLAYWRIGHT_SECRET** 
No longer needed. **Please remove it from the CI secrets as well**

**db-deploy` no longer a dep of `web#build** _(Turbo)_ 
When running `test-e2e`, the dependency `db-deploy` was being run twice by **db-reset** and by **web#build**, causing tests to bootstrap very slow. I remove it from web#build, and replace it as a dependency for web#start as the main idea is to prepare the db before starting the server. This will helps us save time for running an unnecessary step twice.

**no more outputs at test-e2e** _(Turbo)_
This can be considered a minor typo. Removing or leaving this won't make any difference. But as a general rule, cached tasks are not using `outputs` so having this key is unnecessary and can lead to confusion. 

**test not ^test** _(Turbo)_
A minor typo that was looking to tun the child unit test of **calcom/web** instead of its own. In this case, we are more interested in running the unit test cases of calcom/web itself as a dep to e2e. I have also added **^test** as a dep of the general test turbo task so that upcoming child dependency unit tests can run as well.

**wrapping test.use() within test.describe()**
The majority of test files are now wrapped under one or several `test.describe()` blocks when needed. Seems that the call to test.use() is scoped globally to the whole suite when it is called in the global scope of the running module. By placing it within a **describe** block, a call to **test.use()** is now scoped to it. This is intensively used to reload our different sessions and simulate the login of different users. The same applies for test hooks like beforeEach, afterEach, beforeAll, afterAll. They are better placed within a **describe** block than outside in the global scope of the module.

**teardown of db mutations**
As we are running all of our tests under a single and shared db, it is quite important to restore the mutations made in the DB back to the initial state (workers=1 helps on this very much). The goal here is to run all of our tests with a more predictable state instead of an unknown one. 

_**#TODO:**_  A teardown function for the recreation of the delete-me user is still needed. This is not preventing the tests to run but it might be good to have it. I left this to anyone else as I personally prefer to use the fixtures approach here.   

**Replace $$ with Locators**
Playwright warns in their documentation about not using $$ and favor **locators** instead for many reasons. In particular, I found that locators have awaiting logic built into them that will prevent tests from failing too early when the element isn't yet rendered. 

**Fixtures**
IMHO, I believe we should focus on fixtures instead of using the seeded DB as our source of data for tests. Fixtures will allow us to grab a regular **user** generator in each test and create random ones based on the params we want. There is no need to teardown anything here ( I first thought it was) because tests can be scoped to users that can be unique or related to a bunch of tests. The initial db-reset will help to start with a clean known state and then tests will be creating objects as they need. The good thing about this approach is that it can allow us to work with as many workers as we want because there will be no risk of coallisions. When I say users, I also say teams or any other sort of entity Cal needs to model on the tests. 

**FINAL NOTES**
This PR is attempting to recover the working state that E2E was having before with the minimal code change impact.
It does not attempt to refactor already committed tests. A refactoring call to clean up our test suite might be good to address.
Some things that could be refactor:

- Redo empty tests that aren't doing anything or not so much like for eg SAML.test.ts.
- Group auth related tests that are currently distributed in a unorganized manner. Add consistency to the structure of test files.
- Address the logic that shouldn't be run in CI like sending Emails, mocking Stripe. Use **NEXT_PUBLIC_IS_E2E** and **CI** envs.
- Refactor test documentation for contributors.
- Add Fixtures and untight tests using the same data. Raise up the number of workers after this.      

Fixes #2175

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

It can be test in two ways:
- Local Dev Server: `yarn dx && yarn workspace @calcom/web test-e2e`
- Prod/CI Server: `yarn test-e2e`

_The only requirement is that port 3000 needs to be free to be used by the server (local or CI). 
Despite next will automatically assign a free port found on top of 3000 there are some hard code fixes in the CI server (not mine) that requires it to run on port 3000._

### CODE REVIEW

File: apps/web/playwright/change-password.test.ts [hunk 1]
```diff
@@ -1,28 +1,30 @@
 import { expect, test } from "@playwright/test";
 
-// Using logged in state from globalSteup
-test.use({ storageState: "playwright/artifacts/proStorageState.json" });
+test.describe("Change Psssword Test", () => {
```
Comment (zomars) [inline apps/web/playwright/change-password.test.ts:3]:
```suggestion
test.describe("Change Password Test", () => {
```

File: apps/web/playwright/auth/forgot-password.test.ts [hunk 2]
```diff
@@ -3,12 +3,7 @@ import { expect, test } from "@playwright/test";
 test("Can reset forgotten password", async ({ browser }) => {
   test.fixme(true, "TODO: This test is failing randomly, disabled for now");
   // Create a new incognito browser context
-  const context = await browser.newContext({
-    extraHTTPHeaders: {
-      // Only needed for bypassing emails while testing
-      "X-Playwright-Secret": process.env.PLAYWRIGHT_SECRET || "",
-    },
-  });
+  const context = await browser.newContext();
```
[the remark this answers is not in the document]
Comment (denik1981) [reply apps/web/playwright/auth/forgot-password.test.ts:6]: Yes, indeed. Context is not needed anymore. Will update the modification

File: packages/prisma/package.json [hunk 3]
```diff
@@ -14,7 +14,7 @@
     "db-studio": "yarn prisma studio",
     "db-up": "docker-compose up -d",
     "deploy": "run-s build db-deploy",
-    "dx": "yarn db-setup",
+    "dx": "yarn db-reset",
```
Comment (zomars) [inline packages/prisma/package.json:17]: I don't think this is desired behavior for local development.

**Review by zomars — COMMENTED** [review]
Great improvements! Thank you for this. Just left some minor nitpicks. Also `yarn dx` shouldn't reset on every run since it's being used constantly in local development. We don't want to reset the DB on every run.

**Review by zomars — APPROVED** [review]

**Comment by zomars** [comment]
Huge improvements! Thank you for this! 🙏🏽

**Comment by denik1981** [comment]
Thank you so much @zomars for all the comments! It is always so handy to have an extra pair of eyes.

I should have asked about `yarn dx` before because it has nothing related to his issue but more to personal convenience.
The way I think about `dx` is to a handy and unique command that will let me start in dev quickly. It sets up my DB, seeds it  and starts the dev server, all-in-one. 

`db-reset` is only adding a `docker-compose down` step to `db-setup` which actually starts with `db-up` so they seem linked together. The problem with running just `db-setup` only is that the current seed script isn't clearing out the DB when it starts so running `yarn dx` twice will mean attempting to reseed the DB twice with the same data. 

What I usually do in my workflow is run `dx` once at first and then `yarn dev`. But having `db-reset` included within `dx` it becomes handy when I want to start fresh again or if something strange happens with my connection to the postgres docker instance.

I know this is very tight to a personal preference. I will submit a PR with the modification to roll this back so that the team can still enjoy dx as before. I will also update the README to include the `db-reset` step before `dx` when running e2e tests.   

Thank for reviewing the code!

**Pull request merged**